encoding/ssz/detect/configfork.go

```diff
@@ -0,0 +1,178 @@
+package detect
+
+import (
+	"fmt"
+
+	"github.com/prysmaticlabs/prysm/encoding/bytesutil"
+	"github.com/prysmaticlabs/prysm/network/forks"
+
+	ssz "github.com/ferranbt/fastssz"
+	"github.com/pkg/errors"
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state"
+	v1 "github.com/prysmaticlabs/prysm/beacon-chain/state/v1"
+	v2 "github.com/prysmaticlabs/prysm/beacon-chain/state/v2"
+	v3 "github.com/prysmaticlabs/prysm/beacon-chain/state/v3"
+	fieldparams "github.com/prysmaticlabs/prysm/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/config/params"
+	ethpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1/block"
+	"github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1/wrapper"
+	"github.com/prysmaticlabs/prysm/runtime/version"
+	"github.com/prysmaticlabs/prysm/time/slots"
+)
+
+// VersionedUnmarshaler represents the intersection of Configuration (eg mainnet, testnet) and Fork (eg phase0, altair).
+// Using a detected VersionedUnmarshaler, a BeaconState or SignedBeaconBlock can be correctly unmarshaled without the need to
+// hard code a concrete type in paths where only the marshaled bytes, or marshaled bytes and a version, are available.
+type VersionedUnmarshaler struct {
+	Config *params.BeaconChainConfig
+	// Fork aligns with the fork names in config/params/values.go
+	Fork    int
+	Version [fieldparams.VersionLength]byte
```

Comment:
I am not entirely sure why we need the `Fork` field here. Isn't a combination of `Config` and `Version` unique? Also I think `Version` is not a descriptive enough name because it suggests it's the version of the unmarshaler itself.

Reply:
It's used in subsequent PRs. One of the use cases is naming the fork, for instance for human readable file names.

Reply:
It could be changed to a method that uses the fork version schedule to rebuild the value, it is just convenient to keep it around.

Reply:
Let's keep it then, but maybe rename `Version` to `ForkVersion`?